ruby/ext/google/protobuf_c/encode_decode.c

```diff
@@ -645,21 +651,19 @@ VALUE Message_decode(VALUE klass, VALUE data) {
 
   const upb_pbdecodermethod* method = msgdef_decodermethod(desc);
   const upb_handlers* h = upb_pbdecodermethod_desthandlers(method);
-  upb_pbdecoder decoder;
-  upb_sink sink;
-  upb_status status = UPB_STATUS_INIT;
+  upb_env e;
+  upb_env_init(&e);
```

Comment:
I would just benchmark your new implementation (as is) against the old one (before this PR) and see if there is any measurable difference. If not, no need to futz with the more complicated code above.